File: backend/app/schemas/runtime_integration.py
"""Integration snippets and topology→runtime mapping for Step 49."""

from __future__ import annotations

from typing import Any
from uuid import UUID

from pydantic import BaseModel, Field


class IntegrationSnippet(BaseModel):
    id: str
    title: str
    language: str
    content: str


class RuntimeMappingRow(BaseModel):
    topology_node_id: UUID | None = None
    topology_node_name: str | None = None
    resource_id: UUID | None = None
    resource_type: str | None = None
    runtime_name: str | None = None
    container_id: str | None = None
    pod_name: str | None = None
    internal_url: str | None = None
    external_url: str | None = None
    namespace_or_network: str | None = None
    status: str | None = None


class DeploymentIntegrationResponse(BaseModel):
    deployment_id: UUID
    topology_id: UUID
    runtime_provider: str
    namespace_or_network: str | None = None
    internal_endpoints: list[dict[str, Any]] = Field(default_factory=list)
    exposed_endpoints: list[dict[str, Any]] = Field(default_factory=list)
    env_vars: dict[str, str] = Field(default_factory=dict)
    connect_your_app: dict[str, Any] = Field(default_factory=dict)
    snippets: list[IntegrationSnippet] = Field(default_factory=list)
    instructions: dict[str, Any] = Field(default_factory=dict)


class DeploymentRuntimeMappingResponse(BaseModel):
    deployment_id: UUID
    topology_id: UUID
    runtime_provider: str
    rows: list[RuntimeMappingRow] = Field(default_factory=list)

File: frontend/src/api/runtimeIntegration.ts
import { apiFetch } from './client';

export type IntegrationSnippet = {
  id: string;
  title: string;
  language: string;
  content: string;
};

export type DeploymentIntegrationResponse = {
  deployment_id: string;
  topology_id: string;
  runtime_provider: string;
  namespace_or_network?: string | null;
  internal_endpoints: Record<string, unknown>[];
  exposed_endpoints: Record<string, unknown>[];
  env_vars: Record<string, string>;
  connect_your_app: Record<string, unknown>;
  snippets: IntegrationSnippet[];
  instructions: Record<string, unknown>;
};

export type RuntimeMappingRow = {
  topology_node_id?: string | null;
  topology_node_name?: string | null;
  resource_id?: string | null;
  resource_type?: string | null;
  runtime_name?: string | null;
  container_id?: string | null;
  pod_name?: string | null;
  internal_url?: string | null;
  external_url?: string | null;
  namespace_or_network?: string | null;
  status?: string | null;
};

export type DeploymentRuntimeMappingResponse = {
  deployment_id: string;
  topology_id: string;
  runtime_provider: string;
  rows: RuntimeMappingRow[];
};

export type IntegrationServiceOutput = {
  name: string;
  runtime_name?: string | null;
  internal_url?: string | null;
  external_url?: string | null;
  preferred_url?: string | null;
  endpoint_scope: 'external' | 'internal_only' | string;
  url_note?: string | null;
  protocol?: string | null;
  port?: number | null;
  recommended_env_var: string;
  env_vars: Record<string, string>;
};

export type IntegrationOutputsBundle = {
  env: string;
  curl: string;
  bash: string;
  python: string;
  javascript: string;
  typescript: string;
  java: string;
  go: string;
  ruby: string;
  php: string;
  csharp: string;
  github_actions: string;
  docker_compose_env: string;
  kubernetes_configmap: string;
};

export type DeploymentIntegrationOutputsResponse = {
  deployment_id: string;
  topology_id: string;
  runtime_provider: string;
  namespace_or_network?: string | null;
  services: IntegrationServiceOutput[];
  outputs: IntegrationOutputsBundle;
  metadata?: Record<string, unknown>;
};

export type AppLanguageKey =
  | 'curl'
  | 'python'
  | 'javascript'
  | 'typescript'
  | 'java'
  | 'go'
  | 'ruby'
  | 'php'
  | 'csharp';

export const APP_LANGUAGE_OPTIONS: { id: AppLanguageKey; label: string }[] = [
  { id: 'curl', label: 'curl' },
  { id: 'python', label: 'Python' },
  { id: 'javascript', label: 'JavaScript' },
  { id: 'typescript', label: 'TypeScript' },
  { id: 'java', label: 'Java' },
  { id: 'go', label: 'Go' },
  { id: 'ruby', label: 'Ruby' },
  { id: 'php', label: 'PHP' },
  { id: 'csharp', label: 'C#' },
];

export function fetchDeploymentIntegration(deploymentId: string) {
  return apiFetch<DeploymentIntegrationResponse>(`/deployments/${deploymentId}/runtime/integration`);
}

export function fetchDeploymentRuntimeMapping(deploymentId: string) {
  return apiFetch<DeploymentRuntimeMappingResponse>(`/deployments/${deploymentId}/runtime/mapping`);
}

export function fetchDeploymentIntegrationOutputs(deploymentId: string) {
  return apiFetch<DeploymentIntegrationOutputsResponse>(
    `/deployments/${deploymentId}/integration-outputs`,
  );
}
